dynamic-ssl/src/main/java/org/wildfly/security/dynamic/ssl/DynamicSSLContextSpiImpl.java

```diff
@@ -0,0 +1,146 @@
+            dynamicSSLContextImpl = dynamicSSLContextSPIImpl;
+        } else {
+            Iterator<DynamicSSLContextSPI> dynamicSSLContextSPIIterator = ServiceLoader.load(DynamicSSLContextSPI.class).iterator();
+            if (dynamicSSLContextSPIIterator.hasNext()) {
+                dynamicSSLContextImpl = dynamicSSLContextSPIIterator.next();
+            } else {
+                dynamicSSLContextImpl = null;
+            }
+        }
+    }
+
+    private SSLContext getConfiguredDefaultSSLContext() {
+        try {
+            if (dynamicSSLContextImpl != null) {
+                SSLContext configuredDefault = dynamicSSLContextImpl.getConfiguredDefault();
+                if (configuredDefault != null) {
+                    return configuredDefault;
+                }
+            }
+            return SSLContext.getDefault();
+        } catch (NoSuchAlgorithmException | DynamicSSLContextException e) {
+            throw new IllegalStateException(ElytronMessages.log.cannotObtainConfiguredDefaultSSLContext());
+        }
+    }
+
+    @Override
+    protected void engineInit(KeyManager[] keyManagers, TrustManager[] trustManagers, SecureRandom secureRandom) {
+        // ignore
+    }
+
+    @Override
+    protected SSLSocketFactory engineGetSocketFactory() {
+        if (dynamicSSLContextImpl == null) {
+                return this.getConfiguredDefaultSSLContext().getSocketFactory();
+        }
+        if (sslSocketFactory == null) {
+            synchronized (this) {
+                if (sslSocketFactory == null) {
+                    sslSocketFactory = new DynamicSslSocketFactory(this.getConfiguredDefaultSSLContext().getSocketFactory(), dynamicSSLContextImpl);
+                }
+            }
+        }
+        return sslSocketFactory;
+    }
+
+    @Override
+    protected SSLServerSocketFactory engineGetServerSocketFactory() {
+        return this.getConfiguredDefaultSSLContext().getServerSocketFactory();
+    }
+
+    @Override
+    protected SSLEngine engineCreateSSLEngine() {
+            return this.getConfiguredDefaultSSLContext().createSSLEngine();
+    }
+
+    @Override
+    protected SSLEngine engineCreateSSLEngine(String host, int port) throws IllegalStateException {
+        try {
+            if (dynamicSSLContextImpl == null) {
+                return this.getConfiguredDefaultSSLContext().createSSLEngine(host, port);
+            }
+            SSLContext sslContext = dynamicSSLContextImpl
+                    .getSSLContext(new URI(null, null, host, port, null, null, null));
+            if (sslContext == null) {
+                throw new IllegalStateException(ElytronMessages.log.receivedSSLContextFromDynamicSSLContextProviderWasNull());
+            }
+            if (sslContext instanceof DynamicSSLContext && sslContext.getSocketFactory().equals(this.engineGetSocketFactory())) {
+                throw new IllegalStateException(ElytronMessages.log.dynamicSSLContextCreatesLoop());
```

Comment:
Looks like this could be simplified to throw `ElytronMessages.log.dynamicSSLContextCreatesLoop()`

Reply:
Fixed